File: Assets/Scripts/Gameplay/ItemSpawner.cs
using System.Collections.Generic;
using DangerSnake.Core;
using DangerSnake.Items;
using UnityEngine;

namespace DangerSnake.Gameplay
{
    public sealed class ItemSpawner
    {
        readonly GameConfig _config;
        readonly List<FieldItem> _items = new List<FieldItem>(8);
        float _cooldown;

        public IReadOnlyList<FieldItem> Items => _items;

        public ItemSpawner(GameConfig config)
        {
            _config = config;
            _cooldown = config.ItemSpawnInterval * 0.5f;
        }

        public void Reset()
        {
            _items.Clear();
            _cooldown = _config.ItemSpawnInterval * 0.5f;
        }

        public void TickRealtime(float dt, GridModel grid, HashSet<GridPos> occupied)
        {
            _cooldown -= dt;
            if (_cooldown > 0f) return;
            if (_items.Count >= _config.MaxItemsOnField) return;

            if (grid.TryGetEmptyCell(occupied, out var cell))
            {
                var type = (ItemType)Random.Range(0, 4);
                _items.Add(new FieldItem(cell, type));
            }

            _cooldown = _config.ItemSpawnInterval;
        }

        public bool TryPickupAt(GridPos pos, out ItemType type)
        {
            for (var i = 0; i < _items.Count; i++)
            {
                if (_items[i].Position == pos)
                {
                    type = _items[i].Type;
                    _items.RemoveAt(i);
                    return true;
                }
            }

            type = default;
            return false;
        }

        public IEnumerable<GridPos> Positions
        {
            get
            {
                for (var i = 0; i < _items.Count; i++)
                    yield return _items[i].Position;
            }
        }
    }
}

File: DangerSnake/DangerSnake/Gameplay/ItemSpawner.swift
import Foundation

final class ItemSpawner {
    private let config: GameConfig
    private(set) var items: [FieldItem] = []
    private var cooldown: TimeInterval
    private var clock: TimeInterval = 0

    init(config: GameConfig) {
        self.config = config
        self.cooldown = config.itemSpawnInterval * 0.5
    }

    func reset() {
        items.removeAll(keepingCapacity: true)
        cooldown = config.itemSpawnInterval * 0.5
        clock = 0
    }

    var positions: [GridPos] {
        items.map(\.position)
    }

    var greenApples: [FieldItem] {
        items.filter { $0.type == .greenApple }
    }

    var swords: [FieldItem] {
        items.filter { $0.type == .sword }
    }

    func nearestGreenApple(from pos: GridPos) -> FieldItem? {
        greenApples.min { a, b in
            pos.manhattan(a.position) < pos.manhattan(b.position)
        }
    }

    func nearestSword(from pos: GridPos) -> FieldItem? {
        swords.min { a, b in
            pos.manhattan(a.position) < pos.manhattan(b.position)
        }
    }

    /// Advances spawn cooldown and removes expired swords. Returns `true` if a sword vanished by TTL.
    @discardableResult
    func tickRealtime(dt: TimeInterval, grid: GridModel, occupied: Set<GridPos>) -> Bool {
        clock += dt
        let swordVanished = removeExpired()

        cooldown -= dt
        guard cooldown <= 0 else { return swordVanished }
        guard items.count < config.maxItemsOnField else {
            cooldown = config.itemSpawnInterval
            return swordVanished
        }

        if let cell = grid.randomEmptyCell(occupied: occupied) {
            let type = Self.randomSpawnType(greenWeight: config.greenAppleSpawnWeight)
            let expiresAt: TimeInterval? = type == .sword
                ? clock + config.swordLifetimeDuration
                : nil
            items.append(FieldItem(position: cell, type: type, expiresAt: expiresAt))
        }
        cooldown = config.itemSpawnInterval
        return swordVanished
    }

    func tryPickup(at pos: GridPos) -> ItemType? {
        guard let idx = items.firstIndex(where: { $0.position == pos }) else { return nil }
        let type = items[idx].type
        items.remove(at: idx)
        return type
    }

    /// Picks up gear/weapons only; leaves green apples on the field.
    func tryPickupGear(at pos: GridPos) -> ItemType? {
        guard let idx = items.firstIndex(where: { $0.position == pos }) else { return nil }
        let type = items[idx].type
        guard type != .greenApple else { return nil }
        items.remove(at: idx)
        return type
    }

    private func removeExpired() -> Bool {
        let before = items.count
        items.removeAll { item in
            guard let expiresAt = item.expiresAt else { return false }
            return expiresAt <= clock
        }
        return items.count < before
    }

    private static func randomSpawnType(greenWeight: Double) -> ItemType {
        if Double.random(in: 0...1) < greenWeight {
            return .greenApple
        }
        let gear: [ItemType] = [.shield, .bomb, .sword, .boomerang]
        return gear.randomElement() ?? .shield
    }
}
